### ISSUE TITLE
"State" is not passed to promotion action `onActivate` function

### ISSUE THREAD

**Issue by michaelbromley** [vendure-ecommerce/vendure#2944]
**Is your feature request related to a problem? Please describe.**
Requirement: a promotion like "for every N of Product A, add free gift B"

Currently this is not possible because the value of N (which would be calculated by the PromotionCondition and passed as [state](https://docs.vendure.io/guides/core-concepts/promotions/#dependency-relationships) to the PromotionAction) is not available in the side-effect API methods `onActivate` and `onDeactivate`. Thus there is no way to know how many free gifts to add.

**Describe the solution you'd like**
Add a `state` arg to those functions, passing in the data object from the dependent promo condition(s).

**Describe alternatives you've considered**
Re-implement the logic in the action to calculate the number to add, but this is bad for efficiency and maintenance.

**Comment by dlhck** [comment]
Note: This issue needs to be expanded to generally think about the current implementation of promotions. Therefore more research needed.

**Comment by mschipperheyn** [comment]
Yeah, I'm currently working around this by passing orderLineIds that pass the `test` on a custom ctx key. We work extensively with orderLine specific promotions bc we relate subscription to orderLines, and our promotions have "payNow" and "Future" discount option. Feel free to reach out if you're interested to know how we worked around the limitations. 

Another limitations we work around is the filtering of side effect application by only doing this for "new promotions". This creates a lot of problems for us and we forcibly run ALL coupon side effects when the cart changes.
